cql3/prepare_context.cc

```diff
@@ -98,18 +98,41 @@ void prepare_context::set_bound_variables(const std::vector<shared_ptr<column_id
     _target_columns.resize(bn_size);
 }
 
-prepare_context::function_calls_t& prepare_context::pk_function_calls() {
-    return _pk_fn_calls;
+void prepare_context::clear_pk_function_calls_cache() {
+    for (::shared_ptr<cql3::functions::function_call>& fun_call : _pk_fn_calls) {
+        fun_call->set_id(std::nullopt);
+    }
```

Comment:
I wrote down some pros and cons of both options:

### Option A - shared_ptr<> and global clearing:
Pros:
  * More closely matches the previous behaviour
  * Cache is kept because it might be needed in one specific case (PK restrictions by function call in LWT)
          When it turns out that it is not this one specific case we clear it - there will be no other situations when it's needed so it won't cause trouble
Cons:
  * shared_ptr<option<int8_t>> is ugly
  * introduces unexpected behaviour - clearing in one place clears in the other as well

### Option B - local clearing by traversing expressions
Pros:
  * Nice code
   * No surprising behaviour
   * Not clearing cache somewhere shouldn't cause trouble - the function will just put the value in cache and succesive calls will fetch the result from. I can't think of a scenario where this would be bad.

Cons:
  * It's hard to find all places where cache should be cleared - expression is cloned all the time and kept in various places
  * Each time we are cloning an expression we should remember to clear this new place
  * There might be some hidden scenario where having unwanted cache is bad - multiple function calls will return the same value, which could maybe be bad.

I'm not sure what is best, but I think I'm leaning towards Option A.
I made a commit that would change it to B, it's on the branch [term2-pr3-optionB](https://github.com/cvybhu/scylla/tree/term2-pr3-optionB)
If you think B is better, I will add this commit to this PR.

Reply:
> There might be some hidden scenario where having unwanted cache is bad - multiple function calls will return the same value, which could maybe be bad.

This is a good point. We only want to cache these calls in one place - where we generate the partition key for LWT single-partition operations. Any other place it's wasteful or wrong, or both.

This suggests that instead of always populating the id and then clearing it, we go the opposite route and fill the id just for lwt partition keys (especially true if we isolate the partition key into its own expression). So modification_statement and batch_statement would have a separate call to annotate the partition key generation, if they detect they're a LWT operation. btw batch_statement might have to disallow those non-pure functions entirely in partition key context:

```cql
BEGIN UNLOGGED BATCH
INSERT INTO tab (pk, ck) VALUES (non_pure_function(), 1)
INSERT INTO tab (pk, ck) VALUES (non_pure_function(), 2)
APPLY BATCH
```

Nothing says we're allowed to merge the two calls into one.

> If you think B is better, I will add this commit to this PR.

Let's try to reach consensus first.

Reply:
btw, I'm not saying to change the strategy in this series. We can agree on a direction and implement it afterwards. But it helps to know what that direction is.

Reply:
> ```cql
> BEGIN UNLOGGED BATCH
> INSERT INTO tab (pk, ck) VALUES (non_pure_function(), 1)
> INSERT INTO tab (pk, ck) VALUES (non_pure_function(), 2)
> APPLY BATCH
> ```

I think in this case the two insert statements are separate expressions, so the function_calls inside would get separate cache_ids

Reply:
> This suggests that instead of always populating the id and then clearing it, we go the opposite route and fill the id just for lwt partition keys (especially true if we isolate the partition key into its own expression)

By the time we decide whether the cache should be set, the expression was probably copied a couple of times. We would have to track down all these copies and set the ids there. :/

Maybe we could do something like this:
* During creation (in Cql.g) we set and id for each created function_call
* Then later we can have some global boolean called `use_cache` - and depending on this boolean we would set/get values from cache or not

This would be correct and there wouldn't be any trouble with shared_ptrs

Reply:
> > This suggests that instead of always populating the id and then clearing it, we go the opposite route and fill the id just for lwt partition keys (especially true if we isolate the partition key into its own expression)
> 
> By the time we decide whether the cache should be set, the expression was probably copied a couple of times. We would have to track down all these copies and set the ids there. :/


I don't understand. Why would there be any copies? If we extract the partition key parts into some variable, we only have to compute the cache for that variable.

> 
> Maybe we could do something like this:
> 
> * During creation (in Cql.g) we set and id for each created function_call
> * Then later we can have some global boolean called `use_cache` - and depending on this boolean we would set/get values from cache or not
> 
> This would be correct and there wouldn't be any trouble with shared_ptrs or copied expressions


I don't want to involve the parser in some logic that's related to LWT and modification statement. It's much better if the parser can concentrate on just parsing.

Reply:
> I don't understand. Why would there be any copies? If we extract the partition key parts into some variable, we only have to compute the cache for that variable.

We keep the same expressions in multiple places. By the time we decide whether to clear the cache partition key restrictions have been copied to at least three places, there might be more that I missed:

See clearing cache function from `optionB` branch
https://github.com/cvybhu/scylla/blob/f88c01569b50f75ebebda2d42e44fc39262762b0/cql3/restrictions/statement_restrictions.cc#L1745